File: src/rm_auto_aim/max_entropy_tracker/src/trackers/adaptive_armor_tracker.cpp
// Copyright (C) Max Entropy Tracker. Licensed under the MIT License.
#include "max_entropy_tracker/trackers/adaptive_armor_tracker.hpp"

#include <cmath>
#include <stdexcept>
#include <iostream>

#include "max_entropy_tracker/utils/angle_utils.hpp"

namespace fyt::auto_aim {

AdaptiveArmorTracker::AdaptiveArmorTracker(const UnifiedConfig &config,
                                           double dt,
                                           bool enable_oscillation)
    : BaseTracker(dt),
      config_(config),
      ukf_(config, dt),
      osc_detector_(50, 0.05, 5, 100, enable_oscillation) {}

/* ================================================================ */
/*  Initialize                                                       */
/* ================================================================ */

void AdaptiveArmorTracker::initialize(const std::vector<ObservationData> &obs,
                                      double r1, double r2, double dza) {
  if (obs.empty()) throw std::invalid_argument("At least one observation required");

  const auto &o = obs.front();
  auto [panel_id, center_yaw, _err] =
      panel_associator_.associate_panel(o.yaw, std::nullopt);

  current_panel_id_ = panel_id;
  reference_center_yaw_ = center_yaw;

  ukf_.initialize(obs, r1, r2, dza, panel_id);

  if (o.timestamp.has_value()) {
    current_time_ = o.timestamp.value();
    last_update_time_ = o.timestamp.value();
  }

  transition_to(TrackerState::TRACKING);
  increment_frame();
}

/* ================================================================ */
/*  Predict                                                          */
/* ================================================================ */

void AdaptiveArmorTracker::predict(std::optional<double> target_time) {
  if (!is_initialized()) return;

  double dt = compute_dt(target_time);
  ukf_.predict(dt);

  if (target_time.has_value())
    current_time_ = target_time.value();
  else if (current_time_.has_value())
    current_time_ = current_time_.value() + dt;

  auto [r1, r2] = ukf_.get_radii();
  if (osc_detector_.update(r1, r2)) reset_parameters();

  reference_center_yaw_ = ukf_.get_yaw();
}

/* ================================================================ */
/*  Update                                                           */
/* ================================================================ */

bool AdaptiveArmorTracker::update(const std::vector<ObservationData> &obs) {
  if (!is_initialized() || obs.empty()) {
    handle_observation_loss(config_.tracker.tracking_thres,
                            config_.tracker.lost_thres);
    std::cout << "Observation empty or tracker not initialized, transitioning to "
                 << "state=" << static_cast<int>(state()) << std::endl;
    return false;
  }

  // Auto-predict to observation time
  std::optional<double> obs_time;
  for (const auto &o : obs)
    if (o.timestamp.has_value()) {
      if (!obs_time.has_value() || o.timestamp.value() > obs_time.value())
        obs_time = o.timestamp;
    }

  if (obs_time.has_value() && current_time_.has_value()) {
    double d = obs_time.value() - current_time_.value();
    if (d > min_dt_) predict(obs_time.value());
  }

  handle_observation_received(config_.tracker.tracking_thres);

  bool success = (obs.size() == 1) ? update_single(obs[0])
                                   : update_dual(obs[0], obs[1]);

  if (obs_time.has_value()) update_time(obs_time.value());
  increment_frame();
  if (success) reference_center_yaw_ = ukf_.get_yaw();
  return success;
}

/* ================================================================ */
/*  Single-observation update                                        */
/* ================================================================ */

bool AdaptiveArmorTracker::update_single(const ObservationData &obs,
                                         double override_pos_confidence) {
  std::cout << "Updating with single observation: x=" << obs.x << " y=" << obs.y
            << " z=" << obs.z << " yaw=" << obs.yaw << std::endl;
  auto idx = ukf_.state_idx();

  auto [panel_id, center_yaw, matching_error] =
      panel_associator_.associate_panel(
          obs.yaw, reference_center_yaw_, obs.z, ukf_.x()(idx.Z()),
          obs.x, obs.y, ukf_.x()(idx.X()), ukf_.x()(idx.Y()),
          ukf_.x()(idx.R1()), ukf_.x()(idx.R2()));

  current_panel_id_ = panel_id;
  std::string r_type = PanelAssociator::get_r_type(panel_id);

  auto [h_label, h_conf] = height_identifier_.identify_single(
      obs.z, panel_id, ukf_.x()(idx.Z()), ukf_.x()(idx.DZA()),
      ukf_.is_dza_converged());
  height_label_ = h_label;
  height_confidence_ = h_conf;

  std::string armor_layer;
  if (h_label == HeightLabel::UPPER)
    armor_layer = "upper";
  else if (h_label == HeightLabel::LOWER)
    armor_layer = "lower";
  // else stays empty → UKF will infer

  double pos_conf = (override_pos_confidence >= 0.0)
                        ? override_pos_confidence
                        : compute_position_confidence(armor_layer, h_conf, r_type);

  double panel_angle = panel_id * (M_PI / 2.0);

  return ukf_.update({obs}, {r_type}, {armor_layer}, h_conf, pos_conf,
                     panel_angle);
}

/* ================================================================ */
/*  Dual-observation update                                          */
/* ================================================================ */

bool AdaptiveArmorTracker::update_dual(const ObservationData &obs1,
                                       const ObservationData &obs2) {
  std::cout << "Updating with dual observations:\n"
            << "  obs1: x=" << obs1.x << " y=" << obs1.y << " z=" << obs1.z
            << " yaw=" << obs1.yaw << "\n"
            << "  obs2: x=" << obs2.x << " y=" << obs2.y << " z=" << obs2.z
            << " yaw=" << obs2.yaw << std::endl;

  // First: single update on obs1 with full position confidence
  bool single_ok = update_single(obs1, 1.0);
  if (!single_ok) {
    std::cout << "[adaptive_tracker::update_dual] update_single(obs1) failed\n";
    return false;
  }

  auto idx = ukf_.state_idx();

  auto [pid1, cw1, _e1] = panel_associator_.associate_panel(
      obs1.yaw, reference_center_yaw_, obs1.z, ukf_.x()(idx.Z()),
      obs1.x, obs1.y, ukf_.x()(idx.X()), ukf_.x()(idx.Y()),
      ukf_.x()(idx.R1()), ukf_.x()(idx.R2()));
  auto [pid2, cw2, _e2] = panel_associator_.associate_panel(
      obs2.yaw, reference_center_yaw_, obs2.z, ukf_.x()(idx.Z()),
      obs2.x, obs2.y, ukf_.x()(idx.X()), ukf_.x()(idx.Y()),
      ukf_.x()(idx.R1()), ukf_.x()(idx.R2()));

  std::string rt1 = PanelAssociator::get_r_type(pid1);
  std::string rt2 = PanelAssociator::get_r_type(pid2);

  std::cout << "identify_dual" << std::endl;
  auto [l1, l2, h_conf] = height_identifier_.identify_dual(obs1.z, obs2.z);
  std::cout << "identify_dual" << std::endl;
  height_confidence_ = h_conf;

  bool dual_ok = ukf_.update({obs1, obs2}, {rt1, rt2}, {l1, l2}, h_conf);
  if (!dual_ok) {
    std::cout << "[adaptive_tracker::update_dual] ukf_.update(dual) failed "  \
      "rt1=" << rt1 << " rt2=" << rt2 << " l1=" << l1 << " l2=" << l2 << " h_conf=" << h_conf << std::endl;
  }
  return dual_ok;
}

/* ================================================================ */
/*  Helpers                                                          */
/* ================================================================ */

double AdaptiveArmorTracker::compute_position_confidence(
    const std::string &armor_layer, double height_confidence,
    const std::string &r_type) const {
  if (armor_layer.empty()) return 0.1;

  auto idx = ukf_.state_idx();
  int r_idx = (r_type == "r1") ? idx.R1() : idx.R2();
  double r_std = std::sqrt(ukf_.P()(r_idx, r_idx));

  double param_conf;
  if (r_std < 0.01)
    param_conf = 1.0;
  else if (r_std < 0.05)
    param_conf = 1.0 - (r_std - 0.01) * (0.4 / 0.04);
  else if (r_std < 0.1)
    param_conf = 0.6 - (r_std - 0.05) * (0.3 / 0.05);
  else
    param_conf = 0.3;

  return std::sqrt(height_confidence * param_conf);
}

void AdaptiveArmorTracker::reset_parameters() {
  auto [r1, r2] = osc_detector_.get_reset_values();
  auto idx = ukf_.state_idx();
  ukf_.x()(idx.R1()) = r1;
  ukf_.x()(idx.R2()) = r2;
}

/* ================================================================ */
/*  State queries                                                    */
/* ================================================================ */

Eigen::Vector3d AdaptiveArmorTracker::get_center_position() const {
  return ukf_.get_center_position();
}
double AdaptiveArmorTracker::get_yaw() const { return ukf_.get_yaw(); }
std::pair<double, double> AdaptiveArmorTracker::get_radii() const {
  return ukf_.get_radii();
}
double AdaptiveArmorTracker::get_dza() const { return ukf_.get_dza(); }
int AdaptiveArmorTracker::get_k() const { return ukf_.get_k(); }
double AdaptiveArmorTracker::get_delta() const { return ukf_.get_delta(); }

}  // namespace fyt::auto_aim

File: src/rm_auto_aim/max_entropy_tracker_python/max_entropy_tracker/trackers/adaptive_armor_tracker.py
"""
自适应装甲板跟踪器

组合使用DualRadiusSpinUKF、PanelAssociator、HeightIdentifier等模块
实现完整的装甲板跟踪功能
"""

import numpy as np
from typing import List, Optional, Dict, Any, Tuple
import logging

from .base_tracker import BaseTracker, TrackerState
from ..filters.dual_radius_spin_ukf import DualRadiusSpinUKF, StateIndex
from ..core.config import UnifiedConfig
from ..core.observation import ObservationData
from ..association.panel_associator import PanelAssociator
from ..association.height_identifier import HeightIdentifier, HeightLabel
from ..association.oscillation_detector import OscillationDetector
from ..utils.output_smoother import OutputSmoother, SmoothedOutput

logger = logging.getLogger(__name__)


class AdaptiveArmorTracker(BaseTracker):
    """
    自适应装甲板跟踪器
    
    核心功能:
    1. 装甲板yaw到中心yaw的转换（通过PanelAssociator）
    2. 自动高度层级识别（通过HeightIdentifier）
    3. 支持单观测和双观测更新
    4. 震荡检测和参数重启（通过OscillationDetector）
    
    装甲板布局（4面固定结构）:
    - Panel 0: offset=0°,   r1, lower
    - Panel 1: offset=90°,  r2, upper
    - Panel 2: offset=180°, r1, lower
    - Panel 3: offset=270°, r2, upper
    """
    
    def __init__(
        self,
        config: UnifiedConfig,
        dt: float | None = None,
        enable_oscillation_detection: bool = False
    ):
        """
        初始化跟踪器
        
        Args:
            config: 统一配置
            dt: 时间步长。如果为None，将使用 config.dt
            enable_oscillation_detection: 是否启用震荡检测
        """
        # 使用配置中的基础 dt 作为默认
        eff_dt = dt if dt is not None else getattr(config, 'dt', 0.05)
        super().__init__(eff_dt)
        
        self.config = config
        
        # 创建UKF滤波器
        self.ukf = DualRadiusSpinUKF(config=config, dt=eff_dt)
        
        # 创建数据关联模块
        self.panel_associator = PanelAssociator()
        self.height_identifier = HeightIdentifier()
        self.oscillation_detector = OscillationDetector(enabled=enable_oscillation_detection)
        
        # 跟踪状态
        self._current_panel_id: int = 0
        self._reference_center_yaw: Optional[float] = None
        
        # 高度识别状态
        self._height_label = HeightLabel.UNKNOWN
        self._height_confidence: float = 0.0
        
        # 调试记录
        self._debug_panel_history: List[Tuple] = []
        self._debug_height_history: List[Tuple] = []
        self._debug_dual_obs_history: List[Tuple] = []
        
        # 输出平滑器（非侵入式后处理）
        self._output_smoother = OutputSmoother(config.smoother)
        self._last_smoothed: Optional[SmoothedOutput] = None
        self._last_is_dual_obs: bool = False
        
        logger.info("AdaptiveArmorTracker initialized")
    
    # ==================== 初始化 ====================
    
    def initialize(
        self,
        observations: List[ObservationData],
        r1: float = 0.15,
        r2: float = 0.20,
        dza: float = 0.0,
        **kwargs
    ) -> None:
        """
        初始化跟踪器
        
        Args:
            observations: 初始观测列表（至少1个）
            r1: 初始半径1
            r2: 初始半径2
            dza: 初始装甲板高度差
        """
        if len(observations) == 0:
            raise ValueError("At least one observation required")
        
        obs = observations[0]
        
        # 首帧Panel关联
        panel_id, center_yaw, _ = self.panel_associator.associate_panel(
            armor_yaw=obs.yaw,
            center_yaw_pred=None
        )
        
        self._current_panel_id = panel_id
        self._reference_center_yaw = center_yaw
        
        # 从装甲板位置反推中心位置
        # armor_pos = center + r * unit(armor_yaw)  (armor_yaw为径向方向：中心指向装甲板)
        # => center = armor - r * unit(armor_yaw)
        r = r1 if panel_id % 2 == 0 else r2
        center_x = obs.x - r * np.cos(obs.yaw)
        center_y = obs.y - r * np.sin(obs.yaw)
        
        # 初始化UKF：传入原始装甲板观测以及panel_id，由UKF在内部反推中心
        self.ukf.initialize([obs], r1=r1, r2=r2, dza=dza, panel_id=panel_id)
        
        # 初始化时间（如果观测带有时间戳）
        if obs.timestamp is not None:
            self._current_time = obs.timestamp
            self._last_update_time = obs.timestamp
            logger.info(f"Time initialized to {obs.timestamp:.3f}s")
        
        self._transition_to(TrackerState.TRACKING)  # 修复：初始化后应该是TRACKING状态
        self._increment_frame()
        
        # 初始化输出平滑器
        self._output_smoother.initialize(r1=r1, r2=r2, dza=dza)
        self._last_is_dual_obs = False
        
        logger.info(f"Tracker initialized at panel {panel_id}, center_yaw={np.degrees(center_yaw):.1f}°")
    
    # ==================== 预测步骤 ====================
    
    def predict(self, target_time: Optional[float] = None) -> None:
        """
        预测步骤
        
        Args:
            target_time: 目标时间戳。如果为None，使用默认dt
        """
        if not self.is_initialized:
            return
        
        # 计算实际dt
        dt = self._compute_dt(target_time)
        
        # 记录dt（诊断用）
        self._dt_history.append(dt)
        
        # 使用实际dt预测
        self.ukf.predict(dt)
        
        # 更新内部时间
        if target_time is not None:
            self._current_time = target_time
        elif self._current_time is not None:
            self._current_time += dt
        # 如果两者都为None，则不更新时间（使用fallback dt）
        
        # 震荡检测
        r1, r2 = self.ukf.get_radii()
        if self.oscillation_detector.update(r1, r2):
            self._reset_parameters()
        
        # 更新参考yaw
        self._reference_center_yaw = self.ukf.get_yaw()
        
        # 通知平滑器predict（无观测时保持时间同步）
        self._output_smoother.predict_only(target_time)
    
    # ==================== 更新步骤 ====================
    
    def update(
        self,
        observations: List[ObservationData],
        **kwargs
    ) -> bool:
        """
        更新步骤
        
        根据观测数量自动选择单观测或双观测更新
        自动使用观测的timestamp计算dt并先进行predict
        
        Args:
            observations: 观测列表
            
        Returns:
            是否更新成功
        """
        if not self.is_initialized or len(observations) == 0:
            self._handle_observation_loss(
                self.config.tracker.tracking_thres,
                self.config.tracker.lost_thres
            )
            return False
        
        # === 新增：基于观测时间戳自动predict ===
        # 使用所有观测的最大时间戳作为目标预测时间（以处理双观测时间略有不同的情况）
        timestamps = [o.timestamp for o in observations if o.timestamp is not None]
        obs_time = max(timestamps) if timestamps else None
        if len(timestamps) > 1 and max(timestamps) - min(timestamps) > 0.01:
            # 如果同帧观测的时间差超过10ms，记录警告
            logger.debug(f"Observation timestamps differ: min={min(timestamps):.3f}, max={max(timestamps):.3f}")
        if obs_time is not None and self._current_time is not None:
            # 计算到观测时间的dt
            dt = obs_time - self._current_time
            # 允许微小负偏差以容忍时间同步误差
            if dt > self._min_dt:
                # 先预测到观测时间点
                logger.debug(f"Auto-predict to observation time: dt={dt:.4f}s")
                self.predict(obs_time)
            elif dt < -1e-3:
                # 仅在显著倒退时报警
                logger.warning(f"Observation time {obs_time:.3f}s < current time {self._current_time:.3f}s")
        
        self._handle_observation_received(self.config.tracker.tracking_thres)
        
        if len(observations) == 1:
            success = self._update_single(observations[0])
            self._last_is_dual_obs = False
        else:
            success = self._update_dual(observations[0], observations[1])
            self._last_is_dual_obs = True
        
        # 更新时间（使用选定的最大时间戳）
        if obs_time is not None:
            self._update_time(obs_time)
        
        self._increment_frame()
        
        # 更新参考yaw
        if success:
            self._reference_center_yaw = self.ukf.get_yaw()
            
            # 更新输出平滑器
            self._update_smoother(obs_time)
        
        return success
    
    def _update_single(
        self, 
        obs: ObservationData,
        override_position_confidence: Optional[float] = None
    ) -> bool:
        """
        单观测更新
        
        说明：obs.yaw为装甲板yaw（径向方向：中心指向装甲板）
        
        流程:
        1. Panel关联（armor_yaw→panel_id→center_yaw）
        2. 高度层级识别
        3. 计算置信度
        4. 调用UKF更新
        
        Args:
            obs: 观测数据
            override_position_confidence: 如果提供，将覆盖自动计算的position_confidence
        """
        # Panel关联（根据armor_yaw确定panel_id，并转换为center_yaw）
        panel_id, center_yaw, matching_error = self.panel_associator.associate_panel(
            armor_yaw=obs.yaw,  # 输入：装甲板yaw（径向方向）
            center_yaw_pred=self._reference_center_yaw,
            z_obs=obs.z,
            center_z=self.ukf.x[StateIndex.Z],
            obs_x=obs.x,
            obs_y=obs.y,
            center_x=self.ukf.x[StateIndex.X],
            center_y=self.ukf.x[StateIndex.Y],
            r1=self.ukf.x[StateIndex.R1],
            r2=self.ukf.x[StateIndex.R2]
        )
        
        self._current_panel_id = panel_id
        r_type = self.panel_associator.get_r_type(panel_id)
        
        # 高度层级识别
        height_label, height_confidence = self.height_identifier.identify_single(
            z_obs=obs.z,
            panel_id=panel_id,
            z_mean=self.ukf.x[StateIndex.Z],
            dza=self.ukf.x[StateIndex.DZA],
            dza_converged=self.ukf.is_dza_converged()
        )
        
        self._height_label = height_label
        self._height_confidence = height_confidence
        
        # 转换层级标签
        if height_label == HeightLabel.UPPER:
            armor_layer = 'upper'
        elif height_label == HeightLabel.LOWER:
            armor_layer = 'lower'
        else:
            armor_layer = None
        
        # 计算位置置信度（可以被override覆盖）
        if override_position_confidence is not None:
            position_confidence = override_position_confidence
        else:
            position_confidence = self._compute_position_confidence(
                armor_layer, height_confidence, r_type
            )
        
        # 注意：传递原始装甲板yaw给UKF，而不是center_yaw
        # UKF内部会进行模态选择和delta转换
        # 计算panel_angle供观测模型使用
        panel_angle = panel_id * (np.pi / 2)  # Panel 0,1,2,3 → 0, π/2, π, 3π/2
        
        # 调用UKF更新
        success = self.ukf.update(
            observations=[obs],  # 传递原始观测，包含装甲板yaw
            r_types=[r_type],
            armor_layers=[armor_layer],
            height_confidence=height_confidence,
            position_confidence=position_confidence,
            panel_angle=panel_angle  # 传递panel_angle
        )
        
        # 调试记录
        self._debug_panel_history.append((
            self._frame_count, obs.yaw, panel_id, r_type,
            self._reference_center_yaw, matching_error
        ))
        self._debug_height_history.append((
            self._frame_count, panel_id, armor_layer, height_confidence, obs.z
        ))
        
        return success
    
    def _update_dual(
        self,
        obs1: ObservationData,
        obs2: ObservationData
    ) -> bool:
        """
        双观测更新
        
        策略（与demo1一致）：
        1. 先用第一个观测进行单观测更新（包含Yaw更新，使用高position_confidence）
        2. 再用双观测进行几何更新（仅位置和参数，不更新Yaw）
        
        关键：第一次单观测使用高position_confidence (0.9)，因为：
        - 双观测场景下位置信息更可靠
        - 后续双观测更新会进一步修正位置
        - 不应因参数未收敛而过度抑制位置更新
        """
        # 先用第一个观测做单观测更新（会更新Yaw和位置）
        # 在双观测场景中，使用高position_confidence=1.0，完全信任位置观测
        # 因为：1) 双观测提供了强几何约束  2) 后续双观测会进一步修正
        success_single = self._update_single(obs1, override_position_confidence=1.0)
        
        if not success_single:
            return False
        
        # 再进行双观测几何更新
        # Panel关联
        panel_id_1, _, _ = self.panel_associator.associate_panel(
            armor_yaw=obs1.yaw,
            center_yaw_pred=self._reference_center_yaw,
            z_obs=obs1.z,
            center_z=self.ukf.x[StateIndex.Z],
            obs_x=obs1.x,
            obs_y=obs1.y,
            center_x=self.ukf.x[StateIndex.X],
            center_y=self.ukf.x[StateIndex.Y],
            r1=self.ukf.x[StateIndex.R1],
            r2=self.ukf.x[StateIndex.R2]
        )
        
        panel_id_2, _, _ = self.panel_associator.associate_panel(
            armor_yaw=obs2.yaw,
            center_yaw_pred=self._reference_center_yaw,
            z_obs=obs2.z,
            center_z=self.ukf.x[StateIndex.Z],
            obs_x=obs2.x,
            obs_y=obs2.y,
            center_x=self.ukf.x[StateIndex.X],
            center_y=self.ukf.x[StateIndex.Y],
            r1=self.ukf.x[StateIndex.R1],
            r2=self.ukf.x[StateIndex.R2]
        )
        
        r_type_1 = self.panel_associator.get_r_type(panel_id_1)
        r_type_2 = self.panel_associator.get_r_type(panel_id_2)
        
        # 双观测高度识别
        layer_1, layer_2, height_confidence = self.height_identifier.identify_dual(
            z1=obs1.z,
            z2=obs2.z
        )
        
        self._height_confidence = height_confidence
        
        # 数据问题提示：如果面板不是相邻面（panel id不相邻），仅记录warning并继续双观测以便诊断
        diff = abs(panel_id_1 - panel_id_2)
        if diff not in (1, 3):  # 非相邻（例如0-2或1-3）
            logger.warning(
                f"Dual observation: non-adjacent panels detected (panel1={panel_id_1}, panel2={panel_id_2}). "
                "This likely indicates a data-generation or association issue."
            )
            self._debug_dual_obs_history.append((
                self._frame_count, obs1.z, obs2.z, layer_1, layer_2, height_confidence, 'non-adjacent-warning'
            ))
        
        # 不再回退为单观测更新；如果是对角板将由UKF或下游诊断捕获错误信息
        # 这里仅记录额外信息以便之后分析
        # (已在上方根据panel id记录non-adjacent-warning)
        
        # 调用UKF双观测更新（仅更新位置和参数，不更新Yaw）
        success = self.ukf.update(
            observations=[obs1, obs2],
            r_types=[r_type_1, r_type_2],
            armor_layers=[layer_1, layer_2],
            height_confidence=height_confidence
        )
        
        # 调试记录
        self._debug_dual_obs_history.append((
            self._frame_count, obs1.z, obs2.z, layer_1, layer_2, height_confidence
        ))
        
        return success
    
    def _compute_position_confidence(
        self,
        armor_layer: Optional[str],
        height_confidence: float,
        r_type: str
    ) -> float:
        """
        计算位置更新置信度
        
        考虑因素:
        1. 层级识别是否成功
        2. 对应半径的不确定性
        """
        # 层级未知 → 极低置信度
        if armor_layer is None:
            return 0.1
        
        # 获取半径不确定性
        r_idx = StateIndex.R1 if r_type == 'r1' else StateIndex.R2
        r_variance = self.ukf.P[r_idx, r_idx]
        r_std = np.sqrt(r_variance)
        
        # 参数不确定性映射
        if r_std < 0.01:
            param_confidence = 1.0
        elif r_std < 0.05:
            param_confidence = 1.0 - (r_std - 0.01) * (0.4 / 0.04)
        elif r_std < 0.1:
            param_confidence = 0.6 - (r_std - 0.05) * (0.3 / 0.05)
        else:
            param_confidence = 0.3
        
        # 综合置信度
        position_confidence = np.sqrt(height_confidence * param_confidence)
        
        return position_confidence
    
    def _reset_parameters(self):
        """重启参数估计"""
        r1_reset, r2_reset = self.oscillation_detector.get_reset_values()
        
        self.ukf.x[StateIndex.R1] = r1_reset
        self.ukf.x[StateIndex.R2] = r2_reset
        
        logger.info(f"Parameters reset: r1={r1_reset:.3f}, r2={r2_reset:.3f}")
    
    def _update_smoother(self, timestamp: Optional[float] = None):
        """
        在 tracker update 成功后，更新输出平滑器
        
        Args:
            timestamp: 当前时间戳
        """
        # 获取 UKF 原始输出
        center_pos = self.ukf.get_center_position()
        yaw = self.ukf.get_yaw()
        r1, r2 = self.ukf.get_radii()
        dza = self.ukf.get_dza()
        
        # 速度
        idx = StateIndex
        velocity = np.array([
            self.ukf.x[idx.VX],
            self.ukf.x[idx.VY],
            self.ukf.x[idx.VZ],
        ])
        yaw_velocity = self.ukf.x[idx.DELTA_RATE]
        
        # 调用平滑器
        self._last_smoothed = self._output_smoother.smooth(
            center_pos=center_pos,
            yaw=yaw,
            velocity=velocity,
            yaw_velocity=yaw_velocity,
            r1=r1, r2=r2, dza=dza,
            P=self.ukf.P,
            r1_idx=idx.R1, r2_idx=idx.R2, dza_idx=idx.DZA,
            is_dual_obs=self._last_is_dual_obs,
            timestamp=timestamp,
        )
    
    # ==================== 状态获取接口 ====================
    
    def get_center_position(self) -> np.ndarray:
        """获取中心位置 [x, y, z]"""
        return self.ukf.get_center_position()
    
    def get_yaw(self) -> float:
        """获取yaw角"""
        return self.ukf.get_yaw()
    
    def get_radii(self) -> Tuple[float, float]:
        """获取半径估计 (r1, r2)"""
        return self.ukf.get_radii()
    
    def get_dza(self) -> float:
        """获取装甲板高度差"""
        return self.ukf.get_dza()
    
    def get_k(self) -> int:
        """获取离散模态k"""
        return self.ukf.get_k()
    
    def get_delta(self) -> float:
        """获取连续delta角"""
        return self.ukf.get_delta()
    
    def get_panel_id(self) -> int:
        """获取当前panel索引"""
        return self._current_panel_id
    
    def get_height_label(self) -> HeightLabel:
        """获取当前高度标签"""
        return self._height_label
    
    def get_height_confidence(self) -> float:
        """获取高度识别置信度"""
        return self._height_confidence
    
    def get_state(self) -> Dict[str, Any]:
        """获取完整状态"""
        ukf_state = self.ukf.get_state_dict()
        
        state = {
            **ukf_state,
            'tracker_state': self._state.name,
            'frame_count': self._frame_count,
            'current_panel_id': self._current_panel_id,
            'height_label': self._height_label.name,
            'height_confidence': self._height_confidence,
            'reference_center_yaw': self._reference_center_yaw,
        }
        
        # 添加平滑后的输出到状态中
        if self._last_smoothed is not None:
            state['smoothed_center_position'] = self._last_smoothed.center_position.tolist()
            state['smoothed_yaw'] = self._last_smoothed.yaw
            state['smoothed_r1'] = self._last_smoothed.r1
            state['smoothed_r2'] = self._last_smoothed.r2
            state['smoothed_dza'] = self._last_smoothed.dza
            state['structural_converged'] = self._last_smoothed.structural_converged
        
        return state
    
    # ==================== 平滑输出接口 ====================
    
    def get_smoothed_output(self) -> Optional[SmoothedOutput]:
        """
        获取平滑后的完整输出
        
        Returns:
            SmoothedOutput 对象，若未初始化则返回 None
        """
        return self._last_smoothed
    
    def get_smoothed_center_position(self) -> np.ndarray:
        """获取平滑后的中心位置 [x, y, z]"""
        if self._last_smoothed is not None:
            return self._last_smoothed.center_position
        return self.ukf.get_center_position()
    
    def get_smoothed_yaw(self) -> float:
        """获取平滑后的yaw角"""
        if self._last_smoothed is not None:
            return self._last_smoothed.yaw
        return self.ukf.get_yaw()
    
    def get_smoothed_radii(self) -> Tuple[float, float]:
        """获取收敛后的半径估计 (r1, r2)"""
        if self._last_smoothed is not None:
            return (self._last_smoothed.r1, self._last_smoothed.r2)
        return self.ukf.get_radii()
    
    def get_smoothed_dza(self) -> float:
        """获取收敛后的装甲板高度差"""
        if self._last_smoothed is not None:
            return self._last_smoothed.dza
        return self.ukf.get_dza()
    
    def get_smoothed_velocity(self) -> np.ndarray:
        """获取平滑后的速度 [vx, vy, vz]"""
        if self._last_smoothed is not None:
            return self._last_smoothed.velocity
        idx = StateIndex
        return np.array([self.ukf.x[idx.VX], self.ukf.x[idx.VY], self.ukf.x[idx.VZ]])
    
    @property
    def output_smoother(self) -> OutputSmoother:
        """获取输出平滑器实例（用于诊断和参数调整）"""
        return self._output_smoother
    
    # ==================== 调试接口 ====================
    
    def get_panel_statistics(self) -> Dict:
        """获取Panel分配统计"""
        if not self._debug_panel_history:
            return {}
        
        panel_counts = {0: 0, 1: 0, 2: 0, 3: 0}
        errors = []
        
        for _, _, panel_id, _, _, error in self._debug_panel_history:
            panel_counts[panel_id] += 1
            if error > 0:
                errors.append(error)
        
        return {
            'total': len(self._debug_panel_history),
            'panel_distribution': panel_counts,
            'avg_matching_error': np.mean(errors) if errors else 0,
            'max_matching_error': np.max(errors) if errors else 0,
            'z_corrections': self.panel_associator.z_correction_count,
        }
    
    def get_height_statistics(self) -> Dict:
        """获取高度识别统计"""
        if not self._debug_height_history:
            return {}
        
        total = len(self._debug_height_history)
        unknown_count = sum(1 for _, _, layer, _, _ in self._debug_height_history if layer is None)
        
        confidences = [conf for _, _, _, conf, _ in self._debug_height_history]
        
        return {
            'total': total,
            'unknown_count': unknown_count,
            'unknown_rate': unknown_count / total if total > 0 else 0,
            'avg_confidence': np.mean(confidences) if confidences else 0,
            'dual_obs_count': len(self._debug_dual_obs_history),
        }
    
    def clear_debug_history(self):
        """清空调试记录"""
        self._debug_panel_history.clear()
        self._debug_height_history.clear()
        self._debug_dual_obs_history.clear()
